meta/xidel.1
.\" Process this file with
.\" groff -man -Tascii xidel.1
.\"
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.TH XIDEL 1 "JUNE 2016" Linux "User Manuals"
.SH NAME
xidel \- is a command line tool to download and extract data from HTML/XML pages.
.SH SYNOPSIS
.HP 5
The trivial usage is to extract an expression from a webpage like:

.B xidel
http://www.example.org \-\-extract //title
.HP 5
The next important option is \-\-follow to follow links on a page. The following example will print the titles of all pages that are linked from http://www.example.org.

.B xidel
http://www.example.org \-\-follow //a \-\-extract //title 
.SH DESCRIPTION
Xidel supports:
.IP "Extract expressions: there are few different kind of extract expressions" 5
o CSS 3 Selectors    : to extract simple elements
.br
o XPath 3            : to extract values and calculate things with them
.br
o XQuery 3           : to create new documents from the extracted values
.br
o JSONiq             : to work with JSON apis
.br
o Templates          : to extract several expressions in an easy way using a annotated version of the page for pattern-matching
o Multipage templates: i.e. a file that contains templates for several pages
.br
o XPath 2 / XQuery 1 : for legacy queries
.br
.IP "Following:" 5
o HTTP Codes         : Redirections like 30x are automatically followed, while keeping things like cookies
.br
o Links              : It can follow all links on a page as well as some extracted values
.IP "Output formats:"
o Adhoc              : just prints the data in a human readable format
.br
o XML                : encodes the data in XML
.br
o HTML               : encodes the data in HTML
.br
o JSON               : encodes the data in a JSON
.br
o bash/cmd           : exports the data as shell variables
.IP "Connections: HTTP / HTTPS as well as local files or stdin"
.IP "Systems: Windows (using wininet), Linux (using synapse+openssl), Mac (with newest synapse)"
.SH OPTIONS
Call 
.B xidel --help 
to see a list of support command line options.

Call 
.B xidel --usage 
for a full reference.
.SH EXAMPLES \- Basics
.HP 5
.IP "1. Print all urls found by a google search."

.B xidel
http://www.google.de/search?q=test \-\-extract "//a/extract(@href, 'url[?]q=([^&]+)&', 1)[. != '']"

.IP "2. Print the title of all pages found by a google search and download them:"

.B xidel
http://www.google.de/search?q=test \-\-follow "//a/extract(@href, 'url[?]q=([^&]+)&', 1)[. != '']" \-\-extract //title \-\-download '{$host}/'

Generally follow all links on a page and print the titles of the linked pages:

With XPath    : 
.B xidel
http://example.org \-f //a \-e //title

With CSS      : 
.B xidel
http://example.org \-f "css('a')" \-\-css title

With Templates: 
.B xidel
http://example.org \-f "<a>{.}</a>*" \-e "<title>{.}</title>"

.IP "3. Another template example:"

If you have an example.xml file like "<x><foo>ood</foo><bar>IMPORTANT!</bar></x>". You can read the important part like:

.B xidel
example.xml \-e "<x><foo>ood</foo><bar>{.}</bar></x>"

(and this will also check, if the part with the ood is there, and fail otherwise)

.IP "5. Calculate something with XPath:"

.B xidel
\-e "(1 + 2 + 3) * 1000000000000 + 4 + 5 + 6 + 7.000000008"

.IP "6. Print all newest stackoverflow questions with title and url:"

.B xidel
http://stackoverflow.com \-e "<A class='question\-hyperlink'>{title:=text(), url:=@href}</A>*"

.IP "7. Print all reddit comments of an user, with html and url:"

.B xidel
"http://www.reddit.com/user/username/" \-\-extract "<t:loop><div class='usertext\-body'><div>{outer\-xml(.)}</div></div><ul class='flat\-list buttons'><a><t:s>link:=@href</t:s>permalink</a></ul></div></div></t:loop>" \-\-follow "<a rel='nofollow next'>{.}</a>?"

.IP "8. Check if your reddit letter is red:"

Webscraping, combining CSS, XPath, JSONiq and automatically form evaluation:

.B xidel
http://reddit.com \-f "form(css('form.login\-form')[1], {'user': '$your_username', 'passwd': '$your_password'})" \-e "css('#mail')/@title"

Using the Reddit API:

.B xidel
\-d "user=$your_username&passwd=$your_password&api_type=json" https://ssl.reddit.com/api/login \-\-method GET 'http://www.reddit.com/api/me.json' \-e '($json).data.has_mail'

.IP "9. Use XQuery, to create a html table of odd and even numbers:"

.B xidel
\-\-xquery '<table>{for $i in 1 to 1000 return <tr><td>{$i}</td><td>{if ($i mod 2 = 0) then "even" else "odd"}</td></tr>}</table>' \-\-output\-format xml

.IP "10. Export variables to bash"

eval "$(xidel http://site \-e 'title:=//title' \-e 'links:=//a/@href' \-\-output\-format bash)"

This sets the bash variable $title to the title of the page and $links becomes an array of all links there

.SH ENVIRONMENT

Use 
.B XIDEL_OPTIONS
to set global command line options.

.SH AUTHOR
Benito van der Zander, <benito_NOSPAM_benibela.de>, 
http://www.benibela.de

Download link: 
http://sourceforge.net/projects/videlibri/files/Xidel/

You can test it online on  
http://www.videlibri.de/cgi-bin/xidelcgi
or directly by sending a request to the cgi service like
http://www.videlibri.de/cgi-bin/xidelcgi?data=<html><title>foobar</title></html>&extract=//title&raw=true
.SH "SEE ALSO"
.BR wget (1),
.BR curl (1)


changelog
Some important changes:

--------Xidel 0.9.4 --------
(2016-06-08)

* completes the support for XQuery 3.0 and XPath 3.0, e.g. functions format-*, math:* functions
* uses a new underlying regular expression engine, FLRE, which supports Unicode, \p,\c,\i character classes and is supposed to be extremely fast
* stricter error conditions, invalid XPath queries that were previously evaluated are now rejected, for example
  in regular expressions (?i) or \b, (1 + if(..) then ...) instead of (1 + (if(..) then ...)),  gYear("123") instead of gYear("0123")
* this improves the XQuery Test Suite conformance and it passes basically 100% of the XPath 2 tests and over 99.5% of the XQuery 3 tests
* supports the EXPath file module to write or read local files
* custom functions: random($max), random-seed(), x:product($seq)
* new multipage template statements: <try><catch>, <else>
* the error raising HTTP codes are now customizable
* the headers of the last HTTP request can be accessed via $headers
* improved debugging facilities: Evaluation tracing, colored output, improved error messages
* the default data model for primitive types is now XML Schema 1.1 and Unicode 8.0 conformant
* boolean operations are short-cut evaluated
* --quiet/-q was renamed to --silent/-s
* check for XIDEL_OPTIONS environment variable
* various fixes and performance improvements

--------Xidel 0.9--------
(2015-06-28)

* support of more XQuery 3 expressions and functions:
  - anonymous functions, higher order function, function references, function tests
  - new flwor clauses: group by, tumbling and sliding window, count, allowing empty
  - try+catch expressions
  - URI qualified names, e.g. Q{http://www.w3.org/2000/xmlns/}foo instead xmlns:foo
  - fn:unparsed-text, parse-xml, environment-variable, path, round#2, string-join#1, fn:serialize ...
* change output format of variables, type annotations, line separators and stderr-information
* add support for multipart form requests and file uploads via a --form parameter
* Add custom functions pxp:transform  and pxp:request-combine
* fix some issues with non ascii characters in cli arguments
* fix raw data/url after --follow=... arguments and multiple --follow with form requests
* fix some HTML parsing bugs
* fix a large number of XPath/XQuery compatibility bugs.


--------Xidel 0.8.4--------

* implementing all syntax extensions described in my Balisage talk:
  - adding pattern matching in let, for, typeswitch expressions
  - different matching modes for attributes
* support for custom headers on the command line
* can import environment variables
* changed html encoding detection to give content-type highest priority
* fix nested forwarding in recursive calls
* fix crashes/exceptions

--------Xidel 0.8--------

* improved JSONiq extension support
  * updated to the new 1.0.3 standard. Most important: () operator to list object keys / unbox arrays, {| .. |} syntax for object merging, relaxed behaviour on sequences
  * changed my dot notation to behave like the dot operator in JSONiq standalone, i.e. disabled the $var.property syntax, but added $var."property" 
  * fix object with keys containing a dot (.)
  * fix parsing of null in json input
* Use arbitrary precision arithmetic for all numeric operations
* Some minor XPath/XQuery 3 syntax: ! and || operators, and switch statement
* new elements in multipage template: <if>, <choose>, <when>, <otherwise> similar to XSLT, <s> to evaluate short XQuery expressions like in singlepage templates  
* allow template:optional flag on all template elements
* improved html form/url handling: input elements can be passed to pxp:form to create their url encoded post data, uri-decode/uri-encode/uri-combine functions to modify urls, resolve-html function that resolves references in html elements (e.g. @src when passed <img>) to the base uri
* more := shenanigans: $var[] := appends to sequences, $sequence[index] := assigns to sequence elements, $array(index) := assigns to array elements, $var(property) := assigns to object properties. Can be combined, so e.g. $array(index)[] inserts in arrays. Assigning () removes elements
* handle <base>-Element for resolving relative links
* Handle hex encoded password in urls
* other minor improvements and bug fixes


--------Xidel 0.7--------

* Improved handling of extract/follow command line chains:
  Following occurs in depth-first-search
  Brackets [ ] can be used to group arguments together (e.g. [ -f //a -e //title ] to read the titles of all pages linked on the current page, without changing the current page)

* The old json/xml output formats are now called json-wrapped and xml-wrapped
* Added html and xml output formats, which will just print the elements in their format 
  with the respective headers  
* Added bash output format to export the variables to the shell (e.g. eval $(xidel --output-format bash ...))

* Fix a few xml/html parsing and serialization bugs (e.g. added closing tags for <link>)
* Added an option to print elements as html, instead as xml

* Added support for JSONiq: e.g. object syntax {"a": 123}("a"), arrays [10,20,30](1), the JSONiq jn:* functions,
  additional pxp:json() and pxp:serialize-json functions to parse/serialize JSON

* Added outer-html, inner-html functions to the XPath engine which will serialize the element
  in a string

* The syntax extension of variable inclusion in strings like "$foo;" was replaced by prefixed strings x"{$foo}", the regex extraction function pxp:filter was renamed to pxp:extract, to avoid confusion with the XQuery 3 fn:filter function

* Added support for different http request methods (e.g. PUT), basic authentication, allow arbitrary ports

* Added a system function to run arbitrary commands, a $raw variable containing the raw data of the current page


web/xidel.php
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<meta name="description" content="Xidel is a command line tool to download html/xml pages and extract data from them using CSS 3 selectors, XPath 2 expressions or pattern-matching templates." />
<meta name="author" content="Benito van der Zander" />
<meta name="keywords" content="Xidel,  xpath, xpath 2, css, css 3, template, pattern matching, xml, html, download" />
<link rel="stylesheet" type="text/css" href="all.css"/>
<link rel="stylesheet" type="text/css" href="xidel.css"/>
<title>Xidel - HTML/XML data extraction tool </title></head>
<body>
<h1><a name="home">Xidel - HTML/XML data extraction tool</a></h1>

<ul id="navigation">
  <li style="border-left: 1px solid blue;"><a href="#home">Start</a></li>
  <li><a href="#features">Features</a></li>
  <li><a href="#examples">Examples</a></li>
  <li><a href="#downloads">Downloads</a></li>
  <li><a href="#script">Xidelscript</a></li>
  <li><a href="#contact">Contact</a></li>
  
  <!---<li style="right: 1em; position:absolute"><a href="index.html">Deutsche Version</a></li>-->
</ul>

<p>
Xidel is a command line tool to download and extract data from html/xml pages.<br><br>

<h2><div><a name="news">News</a></div></h2>
<dl>
<dt><b>2012-11-06</b>: New release</dt>
<dd style="margin-top:8px; margin-bottom: 8px">The 0.6 release adds XQuery support, the <code>form</code> and <code>match</code> functions, improves the Windows commandline interface, merges the two old cgi services to a single one and fixes several interpreter bugs</dd>
<dt><b>2012-09-05</b>: Initial release</dt>
<dd style="margin-top:8px; margin-bottom: 8px">First release of the VideLibri backend as stand-alone cli tool without VideLibri</dd>
</dl> 
<br>

<h2><div><a name="features">Features</a></div></h2>

It supports:
 
<ul class="xidelfeatures">
<li>Extract expressions:<ul>
  <li>CSS 3 Selectors: to extract simple elements</li>
  <li>XPath 2: to extract values and calculate things with them</li>
  <li>XQuery 1: to  create new documents from the extracted values</li> 
  <li>Templates: to extract several expressions in an easy way using a annotated version of the page for pattern-matching</li>
</ul></li>
<li>Following:<ul>
  <li>HTTP Codes: Redirections like 30x are automatically followed, while keeping things like cookies </li>
  <li>Links: It can follow all links on a page as well as some extracted values </li>
</ul></li>
<li>Output formats:<ul>
  <li>Adhoc: just prints the data in a human readable format</li>
  <li>JSON: encodes the data in a json (as array of all pages of objects with the extracted values as properties ) </li>
  <li>XML: encodes the data similarly in a xml  </li>
</ul></li>
<li>Connections: HTTP / HTTPS as well as local files or stdin</li>
<li>Systems: Windows (using wininet), Linux (using synapse+openssl), Mac (? with synapse, not tested) </li>
</ul>

<h2><div><a name="examples">Examples</a></div></h2>

<ol class="examples">
<li>
Print all urls found by a google search.<br><br>
  <code>xidel http://www.google.de/search?q=test --extract "//a/filter(@href, 'url[?]q=([^&]+)&', 1)[. != '']"</code>
</li>
<li>
Print the title of all pages found by a google search and download them: <br><br>
  <code>xidel http://www.google.de/search?q=test --follow "//a/filter(@href, 'url[?]q=([^&]+)&', 1)[. != '']" --extract //title --download '$host;/'</code>
</li>
<li>
Generally follow all links on a page and print the titles of the linked pages:<ul>
  <li>With CSS: <code>xidel http://example.org -f "css('a')" -e "css('title')"</code> </li>
  <li>With XPath: <code>xidel http://example.org -f //a -e //title</code> </li>
  <li>With Templates: <code>xidel http://example.org -f "&lt;a>{.}&lt;/a>*" -e "&lt;title>{.}&lt;/title>"</code> </li>
</ul></li>
<li>Another template example: <br><br>
   If you have an example.xml file like <code>&lt;x>&lt;foo>ood&lt;/foo>&lt;bar>IMPORTANT!&lt;/bar>&lt;/x></code><br>
   You can read the imporant part like: <code>xidel example.xml -e "&lt;x>&lt;foo>ood&lt;/foo>&lt;bar>{.}&lt;/bar>&lt;/x>"</code><br>
   (and this will also check, if the part with the ood is there, and fail otherwise)
</li>
<li>Calculate something with XPath: <br><br>
  <code>xidel -e "(1 + 2 + 3) * 4 + 5 + 6 + 7"</code>
</li>
<li>Print all newest stackoverflow questions with title and url:<br><br>

   <code>xidel http://stackoverflow.com -e "&lt;A class='question-hyperlink'>{title:=text(), url:=@href}&lt;/A>*"</code>
<li>Print all reddit comments of an user, with html and url:<br><br>
   <code>xidel "http://www.reddit.com/user/username/" --extract "&lt;t:loop>&lt;div class='usertext-body'>&lt;div>{outer-xml(.)}&lt;/div>&lt;/div>&lt;ul class='flat-list buttons'>&lt;a>&lt;t:s>link:=@href&lt;/t:s>permalink&lt;/a>&lt;/ul>&lt;/div>&lt;/div>&lt;/t:loop>" --follow "&lt;a rel='nofollow next'>{.}&lt;/a>?"</code>
</li>
<li>Check if your reddit letter is red; combining CSS, XPath and automatically form evaluation:<br><br>
    <code>xidel http://reddit.com -f "form(css('form.login-form')[1], 'user=$your_username&passwd=$your_password')" -e "css('#mail')/@title"</code>
<li>Use XQuery, to create a html table of odd and even numbers: <br><br>
   Windows: <code>xidel -e "xquery version '1.0'; &lt;table>{for $i in 1 to 1000 return &lt;tr>&lt;td>{$i}&lt;/td>&lt;td>{if ($i mod 2 = 0) then 'even' else 'odd'}&lt;/td>&lt;/tr>}&lt;/table>" --printed-node-format xml </code><br>
   Linux: <code>xidel -e 'xquery version "1.0"; &lt;table>{for $i in 1 to 1000 return &lt;tr>&lt;td>{$i}&lt;/td>&lt;td>{if ($i mod 2 = 0) then "even" else "odd"}&lt;/td>&lt;/tr>}&lt;/table>' --printed-node-format xml</code><br>
   <span style="font-size:75%">(xidel itself supports ' and "-quotes on all platforms, but ' does not escape &lt;&gt; in Windows cmd, and " does not escape $ in the Linux shells)</span>  </li>
</ol>
<br>
You may also want to read the <a href="xidel_readme.txt">readme file of Xidel</a>, the documentation of my <a href="http://benibela.de/documentation/internettools/extendedhtmlparser.THtmlTemplateParser.html">template language</a> and <a href="http://www.benibela.de/documentation/internettools/xquery.TXQueryEngine.html"> XPath 2/XQuery library</a>.



<h2><div><a name="downloads">Downloads</a></div></h2>

The following Xidel downloads are available on the <a href="http://sourceforge.net/projects/videlibri/files/Xidel/Xidel%200.6/">sourceforge download page</a>: <br><br>

<table class="downloadTable">
                        <tr><th>Operating System</th><th>Filename</th><th>Size</th></tr>
                        <tr><td>Windows: 32 Bit</td><td><a href="http://sourceforge.net/projects/videlibri/files/Xidel/Xidel%200.6/xidel-0.6.win32.zip/download">
                        xidel-0.6.win32.zip</a></td><td>394.6 kB</td></tr><tr><td>Universal Linux: 64 Bit</td><td><a href="http://sourceforge.net/projects/videlibri/files/Xidel/Xidel%200.6/xidel-0.6.linux64.tar.gz/download">
                        xidel-0.6.linux64.tar.gz</a></td><td>815.3 kB</td></tr><tr><td>Universal Linux: 32 Bit</td><td><a href="http://sourceforge.net/projects/videlibri/files/Xidel/Xidel%200.6/xidel-0.6.linux32.tar.gz/download">
                        xidel-0.6.linux32.tar.gz</a></td><td>684.4 kB</td></tr><tr><td>Source:</td><td><a href="http://sourceforge.net/projects/videlibri/files/Xidel/Xidel%200.6/xidel-0.6.src.tar.gz/download">
                        xidel-0.6.src.tar.gz</a></td><td>1.1 MB</td></tr><tr><td>Debian: 64 Bit</td><td><a href="http://sourceforge.net/projects/videlibri/files/Xidel/Xidel%200.6/xidel_0.6-1_amd64.deb/download">
                        xidel_0.6-1_amd64.deb</a></td><td>795.7 kB</td></tr><tr><td>Debian: 32 Bit</td><td><a href="http://sourceforge.net/projects/videlibri/files/Xidel/Xidel%200.6/xidel_0.6-1_i386.deb/download">
                        xidel_0.6-1_i386.deb</a></td><td>673.6 kB</td></tr></table>


                        
                        <!-- Table generated by: 
                        xidel http://sourceforge.net/projects/videlibri/files/Xidel/Xidel%200.6/  --extract-kind=xquery  -e 'concat("The following Xidel downloads are available on the <a href=&quot;$url;&quot;>sourceforge download page</a>: <br><br>")' -e 'declare function verboseName($n){ concat ( if (contains($n, "win")) then "Windows: " else if (contains($n, "linux")) then "Universal Linux: " else if (contains($n, ".deb")) then "Debian: " else if (contains($n, "src")) then "Source:" else "", if (contains($n, "32") or contains($n, "386")) then "32 Bit" else if (contains($n, "64"))then "64 Bit" else ""  )   };            
                        <table class="downloadTable">
                        <tr><th>Operating System</th><th>Filename</th><th>Size</th></tr>
                        { for $link in match(<TABLE id="files_list"><t:loop><TR class="file warn"><TH><A class="name">{{link := object(), link.verboseName := verboseName(.), link.a := .}}</A></TH><td/><td>{{link.size := .}}</td></TR></t:loop></TABLE>, /).link 
                          order by $link.verboseName descending 
                          return <tr><td>{$link.verboseName}</td><td><a href="{$link.a/@href}">{$link.a/text()}</a></td><td>{$link.size/text()}</td></tr>}</table>'     --printed-node-format xml -->
<br>Usually you can just extract the zip/deb and call Xidel, or copy it somewhere in your PATH, <br>
because it consists of a single binary without any external dependencies, except the standard system libraries (i.e. Windows API or respectively libc).<br>
However, for https connections on Linux openssl and libcrypto (or even openssl-dev?) are also required.
<br><br>

You can also test it <a href="/cgi-bin/xidelcgi">online on a webpage</a> or directly by sending a request to the cgi service like <a href="/cgi-bin/xidelcgi?data=&lt;html>&lt;title>foobar&lt;/title>&lt;/html>&extract=//title&raw=true" rel="nofollow"><code>http://videlibri.sourceforge.net/cgi-bin/xidelcgi?data=&lt;html>&lt;title>foobar&lt;/title>&lt;/html>&extract=//title&raw=true</code></a>.<br><br>

The source is stored in a <a href="http://videlibri.hg.sourceforge.net/hgweb/videlibri/videlibri/">mercurial repository</a> together with the VideLibri source.<br>
The  program is written in FreePascal/Lazarus, so you need these to compile it. First open the internettools.lpk (in the components/pascal directory) in Lazarus, compile and install it there. On Linux that also requires the synapse library, if you do not have it already installed, you should add the components/pascal/import/synapse directory to the package search path.    Then open xidel.lpi (in the programs/internet/xidel directory),  select possibly your platform/OS in the project settings, and compile it.<br>
It should be possible to compile it on Mac.
<br>

<h2><div><a name="script">Xidelscript</a></div></h2>

<p>There is also a Greasemonkey script to automatically generate templates by selecting the interesting values on a webpage. The script intercepts the selection and marks the elements as shown in the screenshots below: 
<br>
<center><a href="img/reddit.png"><img src="img/reddit_tiny.png" title="Selections that create a template to read the reddit frontpage (extracting title/author/time of each link)" alt="Selections that create a template to read the reddit frontpage (extracting title/author/time of each link)"/></a> 
<a href="img/stackoverflow.png"><img src="img/stackoverflow_tiny.png" title="Selections that create a template to read the newest stackoverflow question. (votes/title/author are extracted for each of the questions) " alt="Selections that create a template to read the newest stackoverflow question. (votes/title/author are extracted for each of the questions) "/></a></center>
<br>
<p>You can find the script in the mercurial repository or on <a href="http://userscripts.org/scripts/show/144991">userscripts.org</a> with a detailed description.

<br><br><br><br>



<h2><div><a id="contact">Contact</a></div></h2>
Autor: Benito van der Zander, <a href="benito_NOSPAM_benibela.de">benito_NOSPAM_benibela.de</a>, <a href="http://www.benibela.de/index_en.html">www.benibela.de</a><br>
<div id="sf-logo"><a href="http://sourceforge.net/projects/videlibri"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=359854&amp;type=1" width="125" height="37" border="0" alt="SourceForge.net Logo" /></a></div>

<!-- Piwik -->
<script type="text/javascript">
var pkBaseURL = (("https:" == document.location.protocol) ? "https://sourceforge.net/apps/piwik/videlibri/" : "http://sourceforge.net/apps/piwik/videlibri/");
document.write(unescape("%3Cscript src='" + pkBaseURL + "piwik.js' type='text/javascript'%3E%3C/script%3E"));
</script><script type="text/javascript">
try {
var piwikTracker = Piwik.getTracker(pkBaseURL + "piwik.php", 1);
piwikTracker.trackPageView();
piwikTracker.enableLinkTracking();
} catch( err ) {}
</script><noscript><p><img src="http://sourceforge.net/apps/piwik/videlibri/piwik.php?idsite=1" style="border:0" alt=""/></p></noscript>
<!-- End Piwik Tag -->
</body>
</html>
